awkward-cpp/awkward/cpp/array/_jagged.cpp

```diff
@@ -1,207 +1,223 @@
-#include <stdexcept>
 #include <pybind11/pybind11.h>
 #include <pybind11/numpy.h>
 #include <cinttypes>
+#include <stdexcept>
 namespace py = pybind11;
 
-py::array_t<std::int64_t> offsets2parents_int64(py::array_t<std::int64_t> offsets) {
-    py::buffer_info offsets_info = offsets.request();
-    if (offsets_info.size <= 0) {
-        throw std::invalid_argument("offsets must have at least one element");
-    }
-    auto offsets_ptr = (std::int64_t*)offsets_info.ptr;
-    
-    size_t parents_length = (size_t)offsets_ptr[offsets_info.size - 1];
-    auto parents = py::array_t<std::int64_t>(parents_length);
-    py::buffer_info parents_info = parents.request();
-    
-    auto parents_ptr = (std::int64_t*)parents_info.ptr;
-    
-    size_t j = 0;
-    size_t k = -1;
-    for (size_t i = 0; i < (size_t)offsets_info.size; i++) {
-        while (j < (size_t)offsets_ptr[i]) {
-            parents_ptr[j] = k;
-            j += 1;
-        }
-        k += 1;
-    }
-    
-    return parents;
-}
+struct JaggedArraySrc {
+public:
 
-py::array_t<std::int32_t> offsets2parents_int32(py::array_t<std::int32_t> offsets) {
-    py::buffer_info offsets_info = offsets.request();
-    if (offsets_info.size <= 0) {
-        throw std::invalid_argument("offsets must have at least one element");
-    }
-    auto offsets_ptr = (std::int32_t*)offsets_info.ptr;
-    
-    size_t parents_length = offsets_ptr[offsets_info.size - 1];
-    auto parents = py::array_t<std::int32_t>(parents_length);
-    py::buffer_info parents_info = parents.request();
-    
-    auto parents_ptr = (std::int32_t*)parents_info.ptr;
-    
-    size_t j = 0;
-    size_t k = -1;
-    for (size_t i = 0; i < (size_t)offsets_info.size; i++) {
-        while (j < (size_t)offsets_ptr[i]) {
-            parents_ptr[j] = k;
-            j += 1;
-        }
-        k += 1;
-    }
-    
-    return parents;
-}
+    template <typename T>
+    static py::array_t<T> offsets2parents(py::array_t<T> offsets) {
+        py::buffer_info offsets_info = offsets.request();
+        if (offsets_info.size <= 0) {
+            throw std::invalid_argument("offsets must have at least one element");
```

Comment:
I believe pybind11 translates the `invalid_argument` exception in C++ to a `ValueError` on the python side, and I can catch it accordingly. I've been trying to find every way possible to keep things on the C++ side instead of the python side, so it doesn't have to be compiled in-time.